Crash on Dry Run

Dry run this code:

```py
# load starlark dependencies
load("http.star", "http")
load("encoding/csv.star", "csv")
load("encoding/json.star", "json")
load("re.star", "re")
load("time.star", "time")

# hit the openweathermap API to get data for Brooklyn, NYC
csvDownloadUrl = "https://api.openweathermap.org/data/2.5/weather?q=Brooklyn,ny,usa&APPID=7c350ee6493701b2c9be3e8ed9c5f857&units=imperial"

raw_json = json.decode(http.get(csvDownloadUrl).body())

# set the body
temp = raw_json['main']['temp']
humidity = raw_json['main']['humidity']

ds = dataset.latest()

# append the new data as a row
ds.body.append([[str(time.now()), temp, humidity]])

# set the dataset body
dataset.commit(ds)
```
Error:

```
2.1cdaaffc.chunk.js:2 TypeError: Cannot read properties of undefined (reading 'length')
    at main.97a6ea3c.chunk.js:1
    at la (main.97a6ea3c.chunk.js:1)
    at Va (main.97a6ea3c.chunk.js:1)
    at ra (2.1cdaaffc.chunk.js:2)
    at qs (2.1cdaaffc.chunk.js:2)
    at Du (2.1cdaaffc.chunk.js:2)
    at Su (2.1cdaaffc.chunk.js:2)
    at Ou (2.1cdaaffc.chunk.js:2)
    at vu (2.1cdaaffc.chunk.js:2)
    at 2.1cdaaffc.chunk.js:2
ss @ 2.1cdaaffc.chunk.js:2
n.callback @ 2.1cdaaffc.chunk.js:2
ho @ 2.1cdaaffc.chunk.js:2
ps @ 2.1cdaaffc.chunk.js:2
ju @ 2.1cdaaffc.chunk.js:2
t.unstable_runWithPriority @ 2.1cdaaffc.chunk.js:2
Hr @ 2.1cdaaffc.chunk.js:2
Lu @ 2.1cdaaffc.chunk.js:2
vu @ 2.1cdaaffc.chunk.js:2
(anonymous) @ 2.1cdaaffc.chunk.js:2
t.unstable_runWithPriority @ 2.1cdaaffc.chunk.js:2
Hr @ 2.1cdaaffc.chunk.js:2
Kr @ 2.1cdaaffc.chunk.js:2
qr @ 2.1cdaaffc.chunk.js:2
mu @ 2.1cdaaffc.chunk.js:2
notify @ 2.1cdaaffc.chunk.js:2
t.notifyNestedSubs @ 2.1cdaaffc.chunk.js:2
t.handleChangeWrapper @ 2.1cdaaffc.chunk.js:2
m @ 2.1cdaaffc.chunk.js:2
dispatch @ VM233:3856
(anonymous) @ main.97a6ea3c.chunk.js:1
main.97a6ea3c.chunk.js:1 error parsing websocket message: TypeError: Cannot read properties of undefined (reading 'length')
```

--- a/src/features/workflow/WorkflowScriptStatus.tsx
+++ b/src/features/workflow/WorkflowScriptStatus.tsx
@@ -4,7 +4,7 @@ import { Run } from '../../qri/run'
 import DatasetCommitInfo from '../../chrome/DatasetCommitInfo'
 import Icon from "../../chrome/Icon"
 import ContentLoader from "react-content-loader"
-import { newVersionInfoFromDataset } from '../../qri/versionInfo'
+import { newVersionInfo, newVersionInfoFromDatasetAndRun, VersionInfo } from '../../qri/versionInfo'
 
 interface WorkflowScriptStatusProps {
   run?: Run
@@ -14,6 +14,11 @@ const WorkflowScriptStatus: React.FC<WorkflowScriptStatusProps> = ({
   run
 }) => {
   const dsPreview = run?.dsPreview
+  let vi: VersionInfo = newVersionInfo({})
+  if (dsPreview) {
+    vi = newVersionInfoFromDatasetAndRun(dsPreview, run)
+    vi.commitTitle = "Dry Run Successful"
+  }
 
   let classes = 'border-solid border-qrigray-200 border text-qrigray-400'
   if (run?.status === 'succeeded') {
@@ -33,7 +38,7 @@ const WorkflowScriptStatus: React.FC<WorkflowScriptStatusProps> = ({
           <rect x="72" y="30" width="25" height="10" rx="1" fill="#D5DADD"/>
         </ContentLoader>
         : dsPreview
-          ? <DatasetCommitInfo item={newVersionInfoFromDataset(dsPreview)} small />
+          ? <DatasetCommitInfo item={vi} small />
           : <>
             <div className='text-xs flex items-center'>
               <Icon icon='playCircle' className='mr-1' size='3xs'/>

--- a/src/qri/run.ts
+++ b/src/qri/run.ts
@@ -1,5 +1,5 @@
 import { EventLogLine, EventLogLineType } from "./eventLog"
-import { Dataset } from './dataset'
+import { Dataset, NewDataset } from './dataset'
 
 export type RunStatus =
   | 'waiting'
@@ -15,7 +15,7 @@ export interface Run {
   status: RunStatus
   startTime?: Date
   stopTime?: Date
-  duration?: string
+  duration?: number // duration of the run in nanoseconds
   initID?: string
   steps: RunStep[]
   dsPreview?: Dataset
@@ -40,7 +40,7 @@ export interface RunStep {
   status: RunStatus
   startTime?: Date
   stopTime?: Date
-  duration?: string
+  duration?: number // duration of the step in nanoseconds
   output?: EventLogLine[]
 }
 
@@ -57,6 +57,14 @@ export function NewRunStep (data: Record<string, any>): RunStep {
   }
 }
 
+function toMilliFromNano (na: number): number {
+  return na / 1000000
+}
+
+function toNanoFromMilli (mi: number): number {
+  return mi * 1000000
+}
+
 export function runAddLogStep (run: Run, line: EventLogLine): Run {
   if (run.id !== line.sessionID) {
     return run
@@ -66,13 +74,16 @@ export function runAddLogStep (run: Run, line: EventLogLine): Run {
     case EventLogLineType.ETTransformStart:
       run.status = 'running'
       run.initID = line.data.initID
-      run.startTime = new Date(line.ts / 1000)
+      run.startTime = new Date(toMilliFromNano(line.ts))
       run.steps = []
       break
     case EventLogLineType.ETTransformStop:
       run.initID = line.data.initID
       run.status = line.data.status || 'failed'
-      run.stopTime = new Date(line.ts / 1000)
+      run.stopTime = new Date(toMilliFromNano(line.ts))
+      if (run.startTime) {
+        run.duration = toNanoFromMilli(run.startTime?.getTime() - run.stopTime?.getTime())
+      }
       break
 
     case EventLogLineType.ETTransformStepStart:
@@ -81,14 +92,17 @@ export function runAddLogStep (run: Run, line: EventLogLine): Run {
       }
       const s = NewRunStep(line.data)
       s.status = 'running'
-      s.startTime = new Date(line.ts / 1000)
+      s.startTime = new Date(line.ts)
       run.steps.push(s)
       break
     case EventLogLineType.ETTransformStepStop:
       const step = lastStep(run)
       if (step) {
-        step.stopTime = new Date(line.ts / 1000)
+        step.stopTime = new Date(line.ts)
         step.status = line.data.status || 'failed'
+        if (step.startTime) {
+          step.duration = toNanoFromMilli(step.stopTime.getTime() - step.startTime.getTime())
+        }
       }
       break
     case EventLogLineType.ETTransformStepSkip:
@@ -110,7 +124,7 @@ export function runAddLogStep (run: Run, line: EventLogLine): Run {
       break
 
     case EventLogLineType.ETDatasetPreview:
-      run.dsPreview = line.data as Dataset
+      run.dsPreview = NewDataset(line.data)
   }
   return run
 }

--- a/src/qri/versionInfo.ts
+++ b/src/qri/versionInfo.ts
@@ -1,6 +1,7 @@
+import { runEndTime } from "../utils/runEndTime"
 import Dataset from "./dataset"
 import { QriRef } from "./ref"
-import { RunStatus } from './run'
+import { Run, RunStatus } from './run'
 // VersionInfo pulls details from a dataset at a specific commit in a version
 // history. It's flat, plain data representation of a dataset meant for listing.
 // VersionInfo is a superset of a Reference, embedding all fields a reference
@@ -140,4 +141,18 @@ export function newVersionInfoFromDataset (ds: Dataset): VersionInfo {
     commitTime: ds.commit?.timestamp,
     runID: ds.commit?.runID
   })
+}
+
+export function newVersionInfoFromDatasetAndRun (ds: Dataset, run: Run): VersionInfo {
+  const vi = newVersionInfoFromDataset(ds)
+  vi.runID = run.id
+  vi.runDuration = run.duration
+  if (run.startTime) {
+    vi.runStart = run.startTime.toString()
+    if (vi.runDuration) {
+      vi.commitTime = runEndTime(vi.runStart, vi.runDuration).toString()
+    }
+  }
+  vi.runStatus = run.status
+  return vi
 }

--- a/src/utils/runEndTime.ts
+++ b/src/utils/runEndTime.ts
@@ -1,5 +1,10 @@
-export function runEndTime (runStart: string, runDuration: number): Date {
-  const runStartTime = new Date(runStart)
+export function runEndTime (runStart: string | Date, runDuration: number): Date {
+  let runStartTime = new Date()
+  if (typeof runStart === 'string') {
+    runStartTime = new Date(runStart)
+  } else {
+    runStartTime = runStart
+  }
   if (!runDuration) { runDuration = 0 }
   return new Date(runStartTime.getTime() + (runDuration / 1000000))
 }

